autotest/ogr/ogr_sql_test.py

```diff
@@ -1551,3 +1551,173 @@ def test_ogr_sql_min_max_string_field(data_ds):
         feat = sql_lyr.GetNextFeature()
         assert feat["MIN_str_field"] == "ab"
         assert feat["MAX_str_field"] == "z"
+
+
+##############################################################################
+# Test SELECT * EXCEPT
+
+# Test some error cases. Some of these could potentially be tolerated
+# in the future.
+@pytest.mark.parametrize(
+    "body",
+    [
+        "(",
+        ")",
+        "()",
+        "(*)",
+        "(EASID, ",
+        "(EASID, DOES_NOT_EXIST)",
+        "(EAS_ID, EAS_ID)",
+        "(EAS_ID,, AREA)",
+    ],
+)
+def test_ogr_sql_select_except_errors(data_ds, body):
+    with gdal.quiet_errors():
+        lyr = data_ds.ExecuteSQL(f"SELECT * EXCEPT {body} FROM poly")
+    assert lyr is None
+
+
+def test_ogr_sql_select_except_attrs(data_ds):
+
+    with data_ds.ExecuteSQL("SELECT * EXCEPT (EAS_ID, PRFEDEA) FROM poly") as sql_lyr:
+        assert sql_lyr.GetLayerDefn().GetFieldCount() == 1
+        assert sql_lyr.GetLayerDefn().GetGeomFieldCount() == 1
+
+
+def test_ogr_sql_select_except_geom(data_ds):
+
+    with data_ds.ExecuteSQL(
+        'SELECT * EXCEPT (EAS_ID, "_ogr_geometry_") FROM poly'
+    ) as sql_lyr:
+        assert sql_lyr.GetLayerDefn().GetFieldCount() == 2
+        assert sql_lyr.GetLayerDefn().GetGeomFieldCount() == 0
+
+
+def test_ogr_sql_select_except_multiple_asterisk_1(data_ds):
+    with data_ds.ExecuteSQL("SELECT * EXCEPT (EAS_ID), * FROM poly") as sql_lyr:
+
+        defn = sql_lyr.GetLayerDefn()
+
+        assert defn.GetFieldCount() == 5
+
+        assert defn.GetFieldDefn(0).GetName() == "AREA"
+        assert defn.GetFieldDefn(1).GetName() == "PRFEDEA"
+        assert defn.GetFieldDefn(2).GetName() == "AREA"
+        assert defn.GetFieldDefn(3).GetName() == "EAS_ID"
+        assert defn.GetFieldDefn(4).GetName() == "PRFEDEA"
+
+        assert sql_lyr.GetLayerDefn().GetGeomFieldCount() == 2
+
+
+def test_ogr_sql_select_except_multiple_asterisk_2(data_ds):
+    with data_ds.ExecuteSQL("SELECT *, * EXCEPT (EAS_ID) FROM poly") as sql_lyr:
+
+        defn = sql_lyr.GetLayerDefn()
+
+        assert defn.GetFieldCount() == 5
+
+        assert defn.GetFieldDefn(0).GetName() == "AREA"
+        assert defn.GetFieldDefn(1).GetName() == "EAS_ID"
+        assert defn.GetFieldDefn(2).GetName() == "PRFEDEA"
+        assert defn.GetFieldDefn(3).GetName() == "AREA"
+        assert defn.GetFieldDefn(4).GetName() == "PRFEDEA"
+
+        assert sql_lyr.GetLayerDefn().GetGeomFieldCount() == 2
+
+
+def test_ogr_sql_select_except_named_geometry():
+
+    mem_ds = ogr.GetDriverByName("Memory").CreateDataSource("my_ds")
+    mem_lyr = mem_ds.CreateLayer("my_layer")
+
+    mem_lyr.GetLayerDefn().GetGeomFieldDefn(0).SetName("named_geom")
```

Comment:
Changing the name (or other property) of a field or geometry field after it has been created is in the unspecified behaviour category
should rather be something along:
```suggestion
    mem_lyr = mem_ds.CreateLayer("my_layer", geom_type = ogr.wkbNone)
    mem_lyr.CreateGeomField(ogr.GeomFieldDefn("named_geom", ogr.wkbUnknown))
```